core/consensus/grandpa/impl/authority_manager_impl.cpp

```diff
@@ -85,1023 +69,204 @@ namespace kagome::consensus::grandpa {
-        if (ancestor_node != nullptr) {
-          auto node = ancestor_node->block == savepoint_block
-                        ? ancestor_node
-                        : ancestor_node->makeDescendant(savepoint_block,
-                                                        IsBlockFinalized{true});
-          auto res = persistent_storage_->put(
-              storage::kAuthorityManagerStateLookupKey(new_savepoint),
-              storage::Buffer(scale::encode(node).value()));
-          if (res.has_error()) {
-            SL_WARN(logger_,
-                    "Can't make savepoint on block {}: {}",
-                    savepoint_block,
-                    hash_res.error());
-            return res.as_failure();
-          }
-          SL_DEBUG(logger_, "Savepoint has made on block {}", savepoint_block);
-        }
-      } else {
-        SL_WARN(logger_,
-                "Can't take hash of savepoint block {}: {}",
-                new_savepoint,
-                hash_res.error());
-      }
-    }
-
-    auto res = persistent_storage_->put(
-        storage::kAuthorityManagerStateLookupKey("last"),
-        storage::Buffer(scale::encode(saving_state_node).value()));
-    if (res.has_error()) {
-      SL_WARN(logger_,
-              "Can't save last state on block {}: {}",
-              saving_state_block,
-              res.error());
-      return res.as_failure();
-    }
-    SL_DEBUG(logger_, "Last state has saved on block {}", saving_state_block);
-
-    last_saved_state_block_ = saving_state_block.number;
-
-    return outcome::success();
-  }
-
   std::optional<std::shared_ptr<const primitives::AuthoritySet>>
   AuthorityManagerImpl::authorities(const primitives::BlockInfo &target_block,
                                     IsBlockFinalized finalized) const {
-    std::unique_lock lock{mutex_};
-
-    auto node = getNode({.block_info = target_block});
-
-    if (node == nullptr) {
+    if (auto r = block_tree_->hasBlockHeader(target_block.hash);
+        not r or not r.value()) {
       return std::nullopt;
     }
-
-    IsBlockFinalized node_in_finalized_chain =
-        node->block == target_block
-            ? (bool)finalized
-            : node->block.number <= block_tree_->getLastFinalized().number;
-
-    auto adjusted_node =
-        node->makeDescendant(target_block, node_in_finalized_chain);
-
-    if (adjusted_node->enabled) {
-      // Original authorities
-      SL_TRACE(logger_,
-               "Pick authority set with id {} for block {}",
-               adjusted_node->authorities->id,
-               target_block);
-      for (auto &authority : adjusted_node->authorities->authorities) {
-        SL_TRACE(
-            logger_, "Authority {}: {}", authority.id.id, authority.weight);
-      }
-      return adjusted_node->authorities;
+    std::unique_lock lock{mutex_};
+    if (auto r = authoritiesOutcome(target_block, finalized.operator bool())) {
+      return std::move(r.value());
+    } else {
```

Comment:
Don't use `else` after `return`

Reply:
Must use `else`, to access local variable `if (auto r)`